cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -1188,6 +1188,24 @@ ArrayKernelExec GenerateTypeAgnosticVarBinaryBase(detail::GetTypeId get_id) {
   }
 }
 
+// similar to GenerateTypeAgnosticPrimitive, but for variable binary and string types
+template <template <typename...> class Generator, typename... Args>
+ArrayKernelExec GenerateTypeAgnosticVarBinary(detail::GetTypeId get_id) {
```

Comment:
This is no longer type agnostic.

Reply:
This is also essentially the same as GenerateVarBinary (we need a way to differentiate the kernels that have a Type0 argument)

Reply:
Yes, I think we should standardize better the naming convention for all GDs (not only for binary/string types). Maybe we can revisit/resolve the naming convention in [ARROW-14134](https://issues.apache.org/jira/browse/ARROW-14134).

The scalar binary kernel exec generators use `EqualTypes` to refer to same input types, so we should use another term of input/output type matching. I think `isomorphic` is the correct terminology to use for the non-type agnostic case.

Another detail is the use of `Var`, I think it should be removed since `Binary` and `String` types imply variable-width. Instead the GDs for `FixedSizeBinary` should use `FixedSize` in its name. This should be resolved in naming convention PR.

For this PR, what about using the following naming conventions for the variable-sized binary/string GDs?
**Note: That the only GD name being changed is the one added in this PR.**
* `GenerateTypeAgnosticVarBinaryBase`
  * no explicit output type -> output type is binary type
  * combines binary/string kernels
* GenerateVarBinaryBase
  * use `Type0`
  * combines binary/string kernels
* GenerateVarBinary
  * use `Type0`
  * independent binary/string kernels
* GenerateIsomorphicVarBinary
  * no explicit output type -> output type same as input type
  * independent binary/string kernels

Reply:
I'd rather continue using VarBinary since Binary would imply to me all binary types, both variable and fixed width.

Isomorphic isn't immediately obvious to me. Maybe the Type0 variants should be named Returning? Since Type0 is usually used as the return type. Also, both GenerateTypeAgnosticVarBinaryBase and GenerateVarBinaryBase are type-agnostic. That said, having both TypeAgnostic and Base is a little redundant. So maybe:

- GenerateTypeAgnosticVarBinary
- GenerateTypeAgnosticVarBinaryReturning
- GenerateVarBinaryReturning
- GenerateVarBinary

though this would be harder to implement in this PR, unfortunately.

Or we could suffix the non-type0 arguments with ToType, e.g. GenerateVarBinaryToVarBinary, GenerateDecimalToDecimal - perhaps a little verbose.

Reply:
Ok, I will keep it simple for this PR and use the `Returning` variant.

I think the using the suffix of the form `ToType` is good. Better explicit and verbose than short and confusing.